docs/design/schemas.md

```diff
@@ -0,0 +1,180 @@
+# Provider, resource, and data source schemas
+
+Providers, resources, and data sources have schemas. This document explores the options for how they relate to each other, and the types provider developers use to define them.
+
+This problem is fairly symmetrical with respect to providers, resources, and data sources, so we choose to focus on resource schemas without loss of generality.
+
+## Relationship of schema to resource
+
+In `helper/schema`, `Resource` is a struct type with a field `Schema`, whose type is `map[string]*Schema`.
+
+In the new framework, `framework.Resource` may be a struct type or interface: see the [Structs and Interfaces](./structs-interfaces.md) design doc. 
+
+Since the schema is defined by the provider code but must be accessible from the framework code, this means that if `framework.Resource` is a struct, the schema can either be a field (`framework.Resource.Schema`), or a method (`framework.Resource.GetSchema()`), i.e. we could have
+```go
+type Resource struct {
+  Schema *tfprotov6.Schema
+}
+```
+
+or
+```go
+type Resource struct {
+  GetSchema func(context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic)
+}
+```
+
+If `framework.Resource` is an interface, the only option is a method (`framework.Resource.GetSchema()`), since the actual resource struct will be a provider-defined type:
+```go
+type Resource interface {
+  GetSchema(ctx context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic)
+}
+```
+
+### Field on struct `framework.Resource.Schema`
+
+This approach is simple and discoverable. The schema is a static piece of data (in conventional providers) known at compile time, so it is idiomatic that it be represented as a field on a struct.
+
+While (depending on the type used to represent the schema - see below) it may not be apparent at compile time if a provider developer has forgotten to fill in a value for `Resource.Schema`, it will be obvious after the most basic of testing if this is the case.
+
+### Method `framework.Resource.GetSchema()`
+
+This approach is also fairly simple and discoverable, and offers the option of returning an error or diag if the schema is not available. This may be a desirable feature for providers written using generated code, where generating the schema could throw an error.
+
+### Both/either
+
+A `Resource` struct could have a `Schema` field, and a `GetSchema` field with func type added later if needed. The framework could use whichever is set, either erroring if both are set or using a documented fallback mechanism to prefer one over the other.
+
+## Types
```

Comment:
The only extra thing I'd hope for here that I didn't see, but don't have guidance on how to achieve, is I hope our design can surface the distinction  between the following types and their capabilities:

* Block
* Attribute
* tftypes/cty value

The reason I'm focused on that is I feel like helper/schema conflates all those things, and it has misled a lot of provider developers about the impacts of their choices, given they don't see whether Optional, Computed, or Sensitive is available for fields depending on what type they end up being.

I don't know how to surface that in this, and a fair response to this comment may be "this feels out of scope for this document", but I think I'm arguing we should consider the distinction between those three things "necessary complexity" that we can't abstract away from the user. (Though I think limiting which of them a user can use is fair game, still.)

Reply:
I added a rather weak closing paragraph naming this tradeoff, but leaving the details to the implementation. Not sure if we should go into more detail here - what do you think?